src/Stan/Pattern.hs

```diff
@@ -0,0 +1,104 @@
+{- |
+Copyright: (c) 2020 Kowainik
+SPDX-License-Identifier: MPL-2.0
+Maintainer: Kowainik <xrom.xkov@gmail.com>
+
+Patterns for types and type search.
+-}
+
+module Stan.Pattern
+    ( -- * Type
+      Pattern (..)
+
+      -- * eDSL
+    , (?)
+    , (|||)
+    , (|->)
+    , (|::)
+
+      -- * Common 'Pattern's
+    , listPattern
+    , nonEmptyPattern
+    , listFunPattern
+    , integerPattern
+    , naturalPattern
+    ) where
+
+import Relude.Extra.Lens ((.~))
+
+import Stan.NameMeta (NameMeta (..), mkBaseMeta, moduleNameL)
+
+
+{- | Query pattern used to search types in HIE AST.
+-}
+data Pattern
+    {- | Argument, type or constructor:
+
+    +---------------------+---------------------------------------------------------------------+
+    | @a@                 | @PatternName (NameMeta ... \"a\") []@                               |
+    +---------------------+---------------------------------------------------------------------+
+    | @[a]@               | @PatternName (NameMeta ... \"[]\") [aPattern]@                      |
+    +---------------------+---------------------------------------------------------------------+
+    | @Either Int String@ | @PatternName (NameMeta ... \"Either\") [intPattern, stringPattern]@ |
+    +---------------------+---------------------------------------------------------------------+
+    -}
+    = PatternName NameMeta [Pattern]
+    -- | Function pattern.
+    | PatternFun Pattern Pattern
+    -- | Type wildcard, matches anything.
+    | PatternAnything
+    -- | Choice between patterns. Should match either of them.
+    | PatternOr Pattern Pattern
+    deriving stock (Show, Eq)
+
+-- | Short operator alias for 'PatternAnything'.
+(?) :: Pattern
+(?) = PatternAnything
+
+-- | Short operator alias for 'PatternOr'.
+infixr 3 |||
+(|||) :: Pattern -> Pattern -> Pattern
+(|||) = PatternOr
+
+-- | Short operator alias for 'PatternFun'.
+infixr 4 |->
+(|->) :: Pattern -> Pattern -> Pattern
+(|->) = PatternFun
+
+-- | Short operator alias for 'PatternName'.
+infix 5 |::
+(|::) :: NameMeta -> [Pattern] -> Pattern
+(|::) = PatternName
+
+-- | 'Pattern' for list @[a]@ or @'String'@.
+listPattern :: Pattern
+listPattern =
+    NameMeta
+        { nameMetaName       = "[]"
+        , nameMetaModuleName = "GHC.Types"
+        , nameMetaPackage    = "ghc-prim"
+        } |:: [ (?) ]
+    |||
+    (mkBaseMeta "String" & moduleNameL .~ "GHC.Base") |:: []
```

Comment:
This may look like operator overdose, but I guess the main goal of implementing eDSL is to be able to write shorter code. A few extra ideas we can do:

1. Move big `NameMeta`s into variables, so it could be just `listNameMeta |:: [ (?) ] ||| ...`
2. Create one more operator for cases where we have an only single type, like
    ```haskell
    listName |:: (?)
   ```

Reply:
I think I like a first option more (for consistency) 🤔